Azure-Samples/Iot-Telemetry-Simulator
# Add option to randomly duplicate messages

## Purpose

* Upstream systems generating events usually offer at-least-once delivery guarantees. Given that there is no mechanism in IoT Hub / Event Hubs for distributed transactions, it is always possible for the client not to receive the response from the server after sending a message, or crashing before persisting the response, and therefore to retry sending an event. This results in message duplicates. When testing event processing solution, it's important to inject such behavior so as to test if/how the system reacts to it, e.g. by deduplication (e.g. https://github.com/Azure-Samples/streaming-at-scale#duplicate-event-handling). Added an option to randomly duplicate a fraction of messages, e.g. setting `DuplicateEvery` to `1000` will duplicate messages with a probability of 1/1000 for each, i.e. on average will duplicate every 1000th message. Note that messages may be duplicated more than once, i.e. duplicated N times with a probability of `1/(DuplicateEvery)^N`

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?


```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
Run with `DuplicateEvery` set to `2` and verify that on average, 50% of messages are sent once, 25% are sent twice etc.
